### DISCUSSION TITLE
Android C++

### DISCUSSION THREAD

**Discussion by gibssa** [google/filament#6039] [Q&A] [answered]
Hi again Filament team,

I was hoping you could clarify or publish a tiny [Filament Android C++ example](https://github.com/google/filament/issues/980#issuecomment-472901783).

~

I've been stuck again for a couple days, this time trying to get ```Engine::create()``` to link on Android, but without the JNI wrapper:
<img width="1115" alt="image" src="https://user-images.githubusercontent.com/1821165/188296807-38cb02c8-c752-4226-bc7b-188712793a40.png">

Goal is:
- use Filament's C++ API (not Kotlin), so that I can reuse my existing Filament C++ iOS code
- just get ```Engine::create()``` to link properly (NDK, arm64-only)

Apparently there's been [prior art](https://github.com/google/filament/issues/980#issuecomment-752912246) on this, but in 2022 this code doesn't compile anymore.

Of course, Filament's [own JNI wrapper](https://github.com/google/filament/tree/main/android) is itself an example of how to build Filament C++ against the Android NDK, but I couldn't get this to work. 

So far I've tried:
 - start with `./android/samples/hello-gltf-viewer` and work backwards until it's a simple "Hello, World!"
   - I moved a couple hundred `#include` lines before I gave up on feasibility of this approach
 - start with fresh NDK project, get "Hello Triangle" working
   - each time I've tried approaching closer and closer to Filament's actual JNI wrapper (your `CMakeLists.txt`, `build.gradle`s, etc), I repeatedly hit the same issue:
     - lots of ```undefined reference to `__aarch64_*'``` missing implementations (as in the screenshot above), what I suspect is `-latomic`

~

I've created a tiny (minimum-reproducible) example of my link error, downloadable here:
[HelloFilamentCpp.zip](https://github.com/google/filament/files/9483823/HelloFilamentCpp.zip)

_Note: This example .zip Android project includes ```./hermetic_include``` and ```./hermetic_libs``` folders, which I've copied straight from `./out/android-release/filament` from Filament 1.26.0 after a fresh `./build.sh -p android release` (I'm attaching headers and libs sort of like the "hermetic Python" strategy). It fits well inside Github's 10MB attachment limit. My tiny example project only targets arm64._

My specific questions are:
 - am I naively just missing an Android CMake toolchain file for Filament?
 - would you consider publishing a simple Filament "Hello Triangle" Android example in C++?

Thanks again team!

**Comment by romainguy** [comment, marked as answer]
It looks like you are not linking against the right runtime libraries. You can see how Filament itself is built here:

https://github.com/google/filament/blob/main/build/toolchain-aarch64-linux-android.cmake

There's nothing special about it save for using static libc++. You can also check build.gradle to see how the JNI wrapper is compiled and linked but again nothing special.

**Reply by gibssa** [reply]
Yup you were right. Thanks for the near-instant response  🙏

**Reply by langxuelang** [reply]
> Yup you were right. Thanks for the near-instant response 🙏

Hi,I meet the same problem。How did you solve it？thank you。

**Reply by gibssa** [reply]
I don't have that version of the code anymore, but if I remember correctly I was missing some of the CMake toolchain config from Filament's "JNI wrapper" CMake project that Romain linked above:
https://github.com/google/filament/blob/main/build/toolchain-aarch64-linux-android.cmake

Once I added in some of their toolchain code, I was able to get Android C++ to compile in CMake.
